RFC-0004-expression-eval.md

```diff
@@ -0,0 +1,80 @@
+# **RFC0 for Presto**
+
+## [Title]
+
+Proposers
+
+* Tim Meehan
+
+## [Related Issues]
+
+* RFC-0003
+
+## Summary
+
+Constant folding is a technique used in compilers to evaluate constant expressions at compile time. This RFC proposes to add constant folding to Presto's expression evaluation engine. This will allow Presto to evaluate constant expressions at query planning time, reducing the amount of work that needs to be done at query execution time.
+
+RFC-0003 proposed a mechanism to add constant folding support via remote function evaluation.  This RFC has a similar goal, but represents an improvement over that approach.
+
+## Background
+
+Unfenced functions are functions which execute within the same process as the underlying evaluation engine.  Typically, this term is used in the context of user-defined functions, and it is mean to contrast with fenced functions, which execute in a separate process.  Unfenced functions are more efficient than fenced functions, but they are also more dangerous, as they can crash the process in which they are running.
+
+Unfenced functions may often be implemented to provide an efficient set of functions that are common to a business unit or group.  They allow the users of Presto to efficiently customize Presto into a different flavor specific to their business needs.
+
+A key feature of unfenced functions is that they can be used to implement constant folding.  This is because the function can be evaluated at query planning time, and the result can be used as a constant in the query plan.  This can be a significant performance improvement, as it can reduce the amount of work that needs to be done at query execution time.  For example, if an unfenced function is evaluated over a partition column, the result can be used to prune partitions before the query is executed.
+
+Currently in Presto C++, there is no constant folding support for functions which are written in C++.  In practice, constant folding works for functions which have a corresponding Java implementation, but not for functions which are implemented in C++.  It works by essentially just using a forked set of functions written in Java to evaluate the constant expressions, and then during execution relying on the C++ functions to evaluate the non-constant expressions.  This has a few drawbacks:
```

Comment:
Another issue is that Java and C++ function have slightly difference semantics (regex, json, upper, lower, date_format, cast(double as varchar), etc.). If a function is evaluated in Java during constant folding and then again in C++ the user may experience a "mix" of semantics within a single query.

Reply:
This is a good point which is another motivation.  I will add it.